packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -842,12 +867,20 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
 		}
 
 		const { reason, ...summarizeOptions } = options;
-		const result = this.trySummarizeOnce(
-			{ summarizeReason: `onDemand/${reason}` },
-			summarizeOptions,
-			resultsBuilder,
-		);
-		return result;
+		if (options.retryOnFailure === true) {
+			this.summarizeOnDemandWithRetries(`onDemand;${reason}`, resultsBuilder).catch(
+				(error) => {
+					resultsBuilder.fail("summarize failed", error);
```

Comment:
This will never do anything, `resultsBuilder.build()` will have already been called and returned.

Reply:
Ok never mind, I missed two things:

1. `resultsBuilder` is passed in, so caller continues to monitor it (awaits the promises)
2. `resultsBuilder.build()` just packs up the promises into a convenient object. I assumed it tore-down the builder and gave some independent result.

So a caller can await the promises returned by this function, and whenever `summarizeOnDemandWithRetries` fails those promises will resolve.

Seeing that `resultsBuilder` is passed into `summarizeOnDemandWithRetries`, it's a little surpising you need this catch block to do anything.  I'll take a closer look there.

Reply:
Ok. As we've discussed before it would be sweet to refactor this stuff around these 3 promises.  It is always hard to follow.  I think this approach makes sense though.

Do your tests cover this path?  Maybe put a breakpoint on this line and make sure it's hit and is proven to work as expected.

Reply:
Yeah, it wasn't straightforward to make it work and I don't like the solution. But this is the best I came up with. Good point about testing it works as expected. I will add that.